invenio/modules/search/walkers/match_unit.py

```diff
@@ -62,16 +63,16 @@ def match_unit(record, p, f=None, m='a', wl=None):
     if m != 'e' and isinstance(p, six.string_types):
         p = re.compile(p)
 
-    if isinstance(record, list):
+    if isinstance(record, MutableSequence):
         return any([match_unit(field, p, f=f, m=m, wl=wl)
                     for field in record])
-    elif isinstance(record, dict):
+    elif isinstance(record, MutableMapping):
         return any([match_unit(field, p, f=f, m=m, wl=wl)
                     for field in record.values()])
 
     if m == 'e':
-        return str(record) == p
-    return p.search(str(record)) is not None
+        return record == p
```

Comment:
I'm not sure about this one, is `p` always a string? ( I mean we are not comparing types here)